File: mf2/rust/mf2-prototype/src/locale_key.rs
use std::collections::BTreeMap;

pub fn canonical_locale_key(locale: &str) -> String {
    locale_parts(locale).join("-")
}

pub fn locale_lookup_chain(locale: &str) -> Vec<String> {
    structural_lookup_chain(&canonical_locale_key(locale))
}

pub fn lookup_locale<'a, T>(
    values: &'a BTreeMap<String, T>,
    locale: &str,
    fallback: &str,
) -> Option<&'a T> {
    for candidate in locale_lookup_chain(locale) {
        if let Some(value) = lookup_canonical_key(values, &candidate) {
            return Some(value);
        }
    }
    lookup_canonical_key(values, &canonical_locale_key(fallback))
}

fn lookup_canonical_key<'a, T>(
    values: &'a BTreeMap<String, T>,
    canonical_key: &str,
) -> Option<&'a T> {
    values
        .iter()
        .find(|(key, _)| canonical_locale_key(key) == canonical_key)
        .map(|(_, value)| value)
}

pub(crate) fn plural_lookup_chain(
    locale: &str,
    parents: &'static [(&'static str, &'static str)],
) -> Vec<String> {
    let mut chain = Vec::new();
    append_plural_lookup_chain(&canonical_locale_key(locale), parents, &mut chain);
    chain
}

fn append_plural_lookup_chain(
    locale: &str,
    parents: &'static [(&'static str, &'static str)],
    chain: &mut Vec<String>,
) {
    let mut current = locale.to_string();
    while !current.is_empty() {
        if chain.iter().any(|candidate| candidate == &current) {
            return;
        }
        chain.push(current.clone());
        if let Some(parent) = parents
            .iter()
            .find(|(child, _)| *child == current)
            .map(|(_, parent)| *parent)
        {
            append_plural_lookup_chain(parent, parents, chain);
        }
        current = structural_parent(&current).unwrap_or_default();
    }
}

fn structural_lookup_chain(locale: &str) -> Vec<String> {
    let parts: Vec<_> = locale.split('-').filter(|part| !part.is_empty()).collect();
    (1..=parts.len())
        .rev()
        .map(|length| parts[..length].join("-"))
        .collect()
}

fn locale_parts(locale: &str) -> Vec<String> {
    locale
        .trim()
        .replace('_', "-")
        .split('-')
        .filter(|part| !part.is_empty())
        .enumerate()
        .take_while(|(_, part)| part.len() != 1)
        .map(|(index, part)| canonical_subtag(index, part))
        .collect()
}

fn canonical_subtag(index: usize, part: &str) -> String {
    if index == 0 {
        return part.to_ascii_lowercase();
    }
    if part.len() == 4 && part.chars().all(|ch| ch.is_ascii_alphabetic()) {
        let mut chars = part.chars();
        let first = chars
            .next()
            .map(|ch| ch.to_ascii_uppercase())
            .unwrap_or_default();
        let rest = chars.as_str().to_ascii_lowercase();
        return format!("{first}{rest}");
    }
    if (part.len() == 2 && part.chars().all(|ch| ch.is_ascii_alphabetic()))
        || (part.len() == 3 && part.chars().all(|ch| ch.is_ascii_digit()))
    {
        return part.to_ascii_uppercase();
    }
    part.to_ascii_lowercase()
}

fn structural_parent(locale: &str) -> Option<String> {
    locale
        .rsplit_once('-')
        .map(|(parent, _)| parent.to_string())
}

File: mf2/python/src/mojito_mf2/_locale_key.py
from __future__ import annotations

from collections.abc import Mapping
from typing import TypeVar


T = TypeVar("T")


def canonical_locale_key(locale: str) -> str:
    return "-".join(_locale_parts(locale))


def locale_lookup_chain(locale: str) -> list[str]:
    return _structural_lookup_chain(canonical_locale_key(locale))


def lookup_locale(
    values: Mapping[str, T],
    locale: str,
    fallback: str = "en",
) -> T | None:
    canonical_values = {canonical_locale_key(key): value for key, value in values.items()}
    for candidate in locale_lookup_chain(locale):
        if candidate in canonical_values:
            return canonical_values[candidate]
    return canonical_values.get(canonical_locale_key(fallback))


def plural_lookup_chain(locale: str, parents: Mapping[str, str]) -> list[str]:
    chain: list[str] = []
    _append_plural_lookup_chain(canonical_locale_key(locale), parents, chain)
    return chain


def _append_plural_lookup_chain(locale: str, parents: Mapping[str, str], chain: list[str]) -> None:
    current = locale
    while current:
        if current in chain:
            return
        chain.append(current)
        parent = parents.get(current)
        if parent is not None:
            _append_plural_lookup_chain(parent, parents, chain)
        current = _structural_parent(current)


def _structural_lookup_chain(locale: str) -> list[str]:
    parts = [part for part in locale.split("-") if part]
    return [
        "-".join(parts[:length])
        for length in range(len(parts), 0, -1)
    ]


def _locale_parts(locale: str) -> list[str]:
    raw_parts = [
        part
        for part in locale.strip().replace("_", "-").split("-")
        if part
    ]
    parts: list[str] = []
    for index, part in enumerate(raw_parts):
        if len(part) == 1:
            break
        parts.append(_canonical_subtag(index, part))
    return parts


def _canonical_subtag(index: int, part: str) -> str:
    if index == 0:
        return part.lower()
    if len(part) == 4 and part.isalpha():
        return part.title()
    if (len(part) == 2 and part.isalpha()) or (len(part) == 3 and part.isdigit()):
        return part.upper()
    return part.lower()


def _structural_parent(locale: str) -> str:
    return locale.rsplit("-", 1)[0] if "-" in locale else ""
